File: platforms/mesh_host/ios/mesh_host/Sources/mesh_host/BleFrameCodec.swift
import Foundation

/// Ordered, bounded BLE segmentation around one already-framed Mesh Link record.
enum BleFrameCodec {
  private static let marker: UInt8 = 0x4d
  private static let start: UInt8 = 1
  private static let end: UInt8 = 2
  private static let header = 4
  private static let maxWire = 4163

  static func split(_ frame: [UInt8], maximumWrite: Int) -> [[UInt8]]? {
    guard (3...maxWire).contains(frame.count) else { return nil }
    let payload = max(1, maximumWrite - header)
    return stride(from: 0, to: frame.count, by: payload).map { offset in
      let count = min(payload, frame.count - offset)
      let flags: UInt8 = (offset == 0 ? start : 0) | (offset + count == frame.count ? end : 0)
      // The lower length byte is deliberately truncated. `UInt8(frame.count)`
      // traps in a release iOS build for every protected record above 255 bytes,
      // which made a normal voice fragment terminate the whole app.
      return [marker, flags, UInt8(frame.count >> 8), UInt8(truncatingIfNeeded: frame.count)] + Array(frame[offset..<(offset + count)])
    }
  }

  final class Assembler {
    private var expected = 0
    private var bytes = [UInt8]()

    func accept(_ fragment: Data) -> [UInt8]? { accept(Array(fragment)) }
    func accept(_ fragment: [UInt8]) -> [UInt8]? {
      guard fragment.count > BleFrameCodec.header, fragment[0] == BleFrameCodec.marker else { reset(); return nil }
      let flags = fragment[1]
      let total = Int(fragment[2]) << 8 | Int(fragment[3])
      guard (3...BleFrameCodec.maxWire).contains(total) else { reset(); return nil }
      if flags & BleFrameCodec.start != 0 { expected = total; bytes = [] }
      guard expected == total, expected != 0, bytes.count + fragment.count - BleFrameCodec.header <= expected else { reset(); return nil }
      bytes += fragment[BleFrameCodec.header...]
      guard flags & BleFrameCodec.end != 0 else { return nil }
      let output = bytes.count == expected ? bytes : nil
      reset()
      return output
    }
    func reset() { expected = 0; bytes = [] }
  }
}

File: platforms/mesh_host/android/src/main/kotlin/com/frazko/mesh_host/BleFrameCodec.kt
package com.frazko.mesh_host

/** Ordered, bounded BLE segmentation around one already-framed Mesh Link record. */
internal object BleFrameCodec {
    private const val marker = 0x4d
    private const val start = 1
    private const val end = 2
    private const val header = 4
    private const val maxWire = 4163

    fun split(frame: ByteArray, maximumWrite: Int): List<ByteArray> {
        require(frame.size in 3..maxWire)
        val payload = (maximumWrite - header).coerceAtLeast(1)
        return frame.asList().chunked(payload).mapIndexed { index, chunk ->
            val flags = (if (index == 0) start else 0) or
                (if ((index + 1) * payload >= frame.size) end else 0)
            byteArrayOf(marker.toByte(), flags.toByte(), (frame.size ushr 8).toByte(), frame.size.toByte()) +
                chunk.toByteArray()
        }
    }

    internal class Assembler {
        private var expected = 0
        private var bytes = ByteArray(0)

        fun accept(fragment: ByteArray): ByteArray? {
            if (fragment.size <= header || fragment[0].toInt() and 255 != marker) { reset(); return null }
            val flags = fragment[1].toInt() and 255
            val total = ((fragment[2].toInt() and 255) shl 8) or (fragment[3].toInt() and 255)
            if (total !in 3..maxWire) { reset(); return null }
            if (flags and start != 0) { expected = total; bytes = ByteArray(0) }
            if (expected != total || expected == 0 || bytes.size + fragment.size - header > expected) { reset(); return null }
            bytes += fragment.copyOfRange(header, fragment.size)
            if (flags and end == 0) return null
            val output = bytes.takeIf { it.size == expected }
            reset()
            return output
        }

        fun reset() { expected = 0; bytes = ByteArray(0) }
    }
}
